src/Http/Routing/test/UnitTests/Internal/MapActionExpressionTreeBuilderTest.cs

```diff
@@ -24,44 +24,185 @@ namespace Microsoft.AspNetCore.Routing.Internal
+                };
             }
+        }
 
-            var requestDelegate = MapActionExpressionTreeBuilder.BuildRequestDelegate((Action)TestAction);
+        [Theory]
+        [MemberData(nameof(NoResult))]
+        public async Task RequestDelegateInvokesAction(Delegate @delegate)
+        {
+            var httpContext = new DefaultHttpContext();
+
+            var requestDelegate = MapActionExpressionTreeBuilder.BuildRequestDelegate(@delegate);
 
-            await requestDelegate(null!);
+            await requestDelegate(httpContext);
 
-            Assert.True(invoked);
+            Assert.True(httpContext.Items["invoked"] as bool?);
         }
 
-        [Fact]
-        public async Task RequestDelegatePopulatesFromRouteParameterBasedOnParameterName()
+        public static IEnumerable<object[]> FromRouteResult
+        {
+            get
+            {
+                void TestAction(HttpContext httpContext, [FromRoute] int value)
+                {
+                    StoreInput(httpContext, value);
+                };
+
+                Task TaskTestAction(HttpContext httpContext, [FromRoute] int value)
+                {
+                    StoreInput(httpContext, value);
+                    return Task.CompletedTask;
+                }
+
+                ValueTask ValueTaskTestAction(HttpContext httpContext, [FromRoute] int value)
+                {
+                    StoreInput(httpContext, value);
+                    return ValueTask.CompletedTask;
+                }
+
+                
+
+                return new List<object[]>
+                {
+                    new object[] { (Action<HttpContext, int>)TestAction },
+                    new object[] { (Func<HttpContext, int, Task>)TaskTestAction },
+                    new object[] { (Func<HttpContext, int, ValueTask>)ValueTaskTestAction },
+                };
+            }
+        }
+        private static void StoreInput(HttpContext httpContext, object value)
+        {
+            httpContext.Items.Add("input", value);
+        }
+
+        [Theory]
+        [MemberData(nameof(FromRouteResult))]
+        public async Task RequestDelegatePopulatesFromRouteParameterBasedOnParameterName(Delegate @delegate)
         {
             const string paramName = "value";
             const int originalRouteParam = 42;
 
-            int? deserializedRouteParam = null;
+            var httpContext = new DefaultHttpContext();
+            httpContext.Request.RouteValues[paramName] = originalRouteParam.ToString(NumberFormatInfo.InvariantInfo);
+
+            var requestDelegate = MapActionExpressionTreeBuilder.BuildRequestDelegate(@delegate);
+
+            await requestDelegate(httpContext);
+
+            Assert.Equal(originalRouteParam, httpContext.Items["value"] as int?);
+        }
 
-            void TestAction([FromRoute] int value)
+        public static IEnumerable<object[]> FromRouteOptionalResult
```

Comment:
Nit: Any particular reason for using local functions for FromRouteResult but not FromRouteOptionalResult?

Reply:
It looks like local functions are not generating the optional parameter and handles then in compile time.
According to Sharplab.io, this code:
```csharp
public void M()
{
    Test();
    void Test(int a = 2)
    {
    }
}
```
will be compiled to this:
```csharp
public void M()
{
    <M>g__Test|0_0(2);
}

[CompilerGenerated]
internal static void <M>g__Test|0_0(int a)
{
}
```

Perhaps this is another feature that could be asked from C# team?

Reply:
Issue filed: https://github.com/dotnet/roslyn/issues/51518

I also requested that lambdas add support for optional parameters as part of the C# 10 lambda improvements: https://github.com/dotnet/csharplang/pull/4451#issuecomment-786843267